src/Aggregate/AttributeApplyMethod.php

```diff
@@ -0,0 +1,73 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Patchlevel\EventSourcing\Aggregate;
+
+use Patchlevel\EventSourcing\Attribute\Apply;
+use Patchlevel\EventSourcing\Attribute\StrictApply;
+use function method_exists;
+
+/**
+ * @psalm-require-extends AggregateRoot
+ */
+trait AttributeApplyMethod
+{
+    private static bool $strictApply = false;
+
+    /**
+     * @var array<class-string<AggregateChanged>, string>|null
+     */
+    private static ?array $map = null;
+
+    protected function apply(AggregateChanged $event): void
+    {
+        $map = self::aggregateChangeMethodMap();
+
+        if (!array_key_exists($event::class, $map)) {
+            if (self::$strictApply) {
+                throw new ApplyAttributeNotFound($this, $event);
+            }
+
+            return;
+        }
+
+        $method = $map[$event::class];
+
+        if (!method_exists($this, $method)) {
+            return;
+        }
+
+        $this->$method($event);
+    }
+
+    private static function aggregateChangeMethodMap(): array
+    {
+        if (self::$map !== null) {
+            return self::$map;
+        }
+
+        $reflector = new \ReflectionClass(self::class);
+        $attributes = $reflector->getAttributes(StrictApply::class);
+
+        if (count($attributes) > 0) {
+            self::$strictApply = true;
+        }
+
+        $methods = $reflector->getMethods();
+
+        $map = [];
+
+        foreach ($methods as $method) {
+            $attributes = $method->getAttributes(Apply::class);
+
+            foreach ($attributes as $attribute) {
+                $instance = $attribute->newInstance();
+
```

Comment:
just as an idea maybe :)
```
$aggregateChangedClass = $instance->aggregateChangedClass();
if (!$aggregateChangedClass) {
    $parameters = $method->getParameters();

    if (count($parameters) !== 1) {
        // throw logic custom exception
    }

    $aggregateChangedClass = (string) $parameters[0]->getType();
    $map[$aggregateChangedClass] = $method->getName();
}
```

Reply:
Yeah, i think about it too, but this is not easy. You have union `Foo|Bar` and Intersection `Foo&Bar` types, so you need cover this too. And in the future, maybe you can write this `Foo|(Bar&Baz)`, so we need to cover this cases too or throw an exception. Make this optional is not a BC break and we can check this in the future, if it make sense. But for now, it think we should say it explicit.